sia/src/objects/upload.rs

```diff
@@ -0,0 +1,424 @@
+                                tasks.push(self.try_upload_sector(host_queue.clone(), sector.as_ref().to_vec()));
+                            }
+                        }
+                    }
+                },
+                _ = tokio::time::sleep(Duration::from_secs(15)) => {
+                    // race another host to prevent slow hosts from stalling uploads
+                    tasks.push(self.try_upload_sector(host_queue.clone(), sector.as_ref().to_vec()));
+                }
+            }
+        }
+    }
+}
+
+pub struct Uploader<D: HostDialer> {
+    inner: Arc<UploaderInner<D>>,
+}
+
+impl<D: HostDialer> Uploader<D>
+where
+    D: HostDialer + 'static,
+    D::Error: From<UploadError>,
+{
+    pub fn new(dialer: D, account_key: PrivateKey, max_inflight: usize) -> Self {
+        let semaphore = Semaphore::new(max_inflight);
+        Uploader {
+            inner: Arc::new(UploaderInner {
+                account_key,
+                dialer,
+                semaphore,
+            }),
+        }
+    }
+
+    /// helper to upload shards. A function that does not
+    /// take `self` is necessary for tokio::spawn
+    async fn try_upload_shards(
+        uploader: Arc<UploaderInner<D>>,
+        shards: Vec<Vec<u8>>,
+    ) -> Result<Vec<Sector>, D::Error>
+    where
+        D::Error: From<UploadError>,
+    {
+        let hosts = HostQueue::new(uploader.dialer.hosts(), 2);
+        let mut futures = Vec::new();
+        for shard in shards {
+            futures.push(uploader.upload_slab_sector(hosts.clone(), shard));
+        }
+
+        try_join_all(futures).await
+    }
+
+    /// Uploads the erasure coded shards. The shards
+    /// should be encrypted by the caller.
+    ///
+    /// [upload] should generally be preferred for simplicity.
+    /// This is primarily useful for environments that have
+    /// special concurrency requirements.
+    pub async fn upload_shards<R: AsyncReadExt + Unpin>(
+        &self,
+        shards: Vec<Vec<u8>>,
+        encryption_key: [u8; 32],
+        data_shards: u8,
+        length: usize,
+    ) -> Result<Option<Slab>, D::Error> {
+        if shards.len() < data_shards as usize {
+            return Err(UploadError::NotEnoughShards(shards.len() as u8, data_shards).into());
+        }
+        let sectors = Self::try_upload_shards(self.inner.clone(), shards).await?;
+        let slab = Slab {
+            encryption_key,
+            min_shards: data_shards,
+            sectors,
+            offset: 0,
+            length,
+        };
+        Ok(Some(slab))
+    }
+
+    /// Reads until EOF and uploads all slabs.
+    /// The data will be erasure coded, encrypted,
+    /// and uploaded using the uploader's parameters.
+    pub async fn upload<R: AsyncReadExt + Unpin>(
+        &self,
+        r: &mut R,
+        encryption_key: [u8; 32],
+        data_shards: u8,
+        parity_shards: u8,
+    ) -> Result<Vec<Slab>, D::Error> {
```

Comment:
If you change the top if this `impl` block to 

```rust
impl<D: HostDialer> Uploader<D>
where
    D: HostDialer + 'static,
    D::Error: From<UploadError> + From<erasure_coding::Error>,
{
```

You don't need `map_err` anymore. It seems like while the trait only specifies that the error needs to be able to turn into an `UploadError`, the actual implementation can restrict it further.